### ISSUE TITLE
[jet] use proper Taylor coefficients during calculation

### ISSUE THREAD

**Issue by thisiscam** [jax-ml/jax#29624]
**Title:** `jax.experimental.jet` should store unscaled Taylor coefficients `a_n`, *not* `n!\,a_n`

In *Evaluating Derivatives* (Griewank & Walther, §13) the Taylor series is
$f(x)=\sum_{n=0}^{\infty} a_n x^n,$
where $a_n$ is called the *n*-th Taylor coefficient.

`jax.experimental.jet` instead propagates the factorial-scaled value
$c_n=n! \cdot a_n$ and calls $c_n$ as the Taylor coefficient in the docs.
While this makes the output of `jet` interpretable as an ordinary n-th order derivative of $f$, it forces extra scaling/un-scaling by \$n!\$, which can overflow or amplify rounding error due to the use of factorial (see #9712);

**Proposal:** propagate the unscaled coefficients \$a\_n\$ internally and (optionally) offer a compatibility flag/wrapper current for the `n!` convention.

Below is a comparison of the exponential rule before and after this change:
```python
# Current jet implementation
def _exp_taylor(primals_in, series_in, **_):
  x, = primals_in
  series, = series_in
  u = [x] + series
  v = [lax.exp(x)] + [None] * len(series)
  for k in range(1,len(v)):
    v[k] = fact(k-1) * sum(_scale(k, j) * v[k-j] * u[j] for j in range(1, k+1))
  primal_out, *series_out = v
  return primal_out, series_out
jet_rules[lax.exp_p] = _exp_taylor
```

```python
# Using "true" Taylor coefficient
def _exp_taylor(primals_in, series_in, **_):
  x, = primals_in
  series, = series_in
  u = [x] + series
  v = [lax.exp(x)] + [None] * len(series)
  for k in range(1, len(v)):
    v[k] = sum(j * v[k - j] * u[j] for j in range(1, k + 1)) / k
  primal_out, *series_out = v
  return primal_out, series_out
```

After this change, it would also offer `jet` the possibility to switch to a `scan`/`vmap` implementation, to avoid all the python for-loop unrolling --- currently, because of these unrolling, realistically, one can only go to up to about 10-th order derivative before the XLA compiler hangs. One can see that the code like `fact(k-1)` makes it awkward, though not impossible, to implement everything without python for-loops.

I have implemented some of these changes locally already. I raised the issue to ask if the maintainers would be interested in a PR? Are there any downside to such an unscaled approach?

**Comment by mattjj** [comment]
Thanks for raising this, @thisiscam !

> I raised the issue to ask if the maintainers would be interested in a PR?

Yes, we would, if you think it's the right step. That is, while I would love to think about and work on `jet` myself, none of us have had the time to think about it for a long while. So if someone (like you!) is interested in working on `jet`, and is guided by actually trying to use it, I would both be grateful for any improvements you can contribute upstream, and would defer to your judgement about what those improvements should look like.

> While this makes the output of jet interpretable as an ordinary n-th order derivative of $f$

IIUC, we can maintain backward compatibility of the `jet` user-facing API if we want to, by
1. changing the internal representation not to do the factorial scaling,
2. making the `jet` API wrapper apply the factorial scaling just as the first/last step, and
3. adding a `jet2` function, or maybe just an option on the `jet` API, so users can access the not-factorially-scaled version if they want.

Since there probably aren't many `jet` users, it's also viable not to maintain backward compatibility. But it can be nice to do if it's low-cost, both in the short and long terms.

> Are there any downside to such an unscaled approach?

I don't know! I don't recall us choosing it for any particular reason. We didn't look at Griewank & Walther carefully, and may have just made a bad choice here. So let's not assume there's a secret good reason we chose the factorially-scaled version, and just go with whatever you think makes the most sense.

In addition to comparing to Griewank & Walther, there's this recent work on jet in PyTorch which we could also check against: https://arxiv.org/abs/2505.13644

Actually, I was hoping to investigate something about that work: they report that their PyTorch code, which **is AFAICT running a different algorithm (IIRC using Faa Di Bruno rules directly, at least for low orders)**, runs faster than `jax.experimental.jet` in their experiments. I don't think they did the apples-to-apples comparison of the same algorithm/rules implemented in both PyTorch and JAX. So that made me wonder: should we switch our rules, perhaps at certain orders? Is the speed difference from the algorithm, or from some pytorch vs jax implementation thing, or what? What else is there to learn from that project? Any chance you're interested in looking into that, @thisiscam ?

**Comment by thisiscam** [comment]
Hi @mattjj ! Thanks for the quick response. I will polish my current code and do a PR.

**Comment by thisiscam** [comment]
>In addition to comparing to Griewank & Walther, there's this recent work on jet in PyTorch which we could also check against: https://arxiv.org/abs/2505.13644
> Actually, I was hoping to investigate something about that work: they report that their PyTorch code, which is AFAICT running a different algorithm (IIRC using Faa Di Bruno rules directly, at least for low orders), runs faster than jax.experimental.jet in their experiments. I don't think they did the apples-to-apples comparison of the same algorithm/rules implemented in both PyTorch and JAX. So that made me wonder: should we switch our rules, perhaps at certain orders? Is the speed difference from the algorithm, or from some pytorch vs jax implementation thing, or what? What else is there to learn from that project? Any chance you're interested in looking into that, @thisiscam ?

Thank you for pointing to this reference! It looks interesting --- I will take some time to understand what is going on. 
I wasn't able to find their code, though. Do you know if those are public somewhere? 

I would be interested to understand their algorithm, and if it could be related to what I am trying to do. 
Nonetheless, if we can find their code, I can for sure do a performance comparison between their algorithm and my jet PR (i.e., with/without the factorials, and with/without Python loops).

**Comment by mattjj** [comment]
> (i.e., with/without the factorials, and with/without Python loops).

The main difference I'm interested in is if the primitive rules are different, e.g. if we should just use the generic rule in #13737 rather than the specialized O(k^2) rules like we do now, at least for smaller `k`.

**Comment by mattjj** [comment]
> if we can find their code

Hmm, it must be in a private repo.

**Comment by thisiscam** [comment]
Could you explain why you'd want a O(k^3) solution if O(k^2) solution exists and is implemented? 
From what I can tell, both the O(k^3) and O(k^2) solutions are implemented via Python loops --- is it that you think the XLA compiler can discover more optimization opportunities with the O(k^3) unrolled solution?

I can understand that, for currently missing jet primitives, that O(k^3) approach looks like a nice (temporary) alternative. 

On a related note, I have been worked on O(k log(k)) FFT-based solution, and I have been falling back to the O(k^2) solution for smaller k.

**Comment by thisiscam** [comment]
@mattjj I have created the above PR. Let me know what you think!

I also have local changes that implements all existing `jet` primitives using jax loop primitives (e.g.  `scan` and `dynamic_slice`) instead of Python loops. These changes assume a proper Taylor coefficient representation, because otherwise the factorially-scaled representation is tedious to work with, so merging this current PR is a good first step to that.

**Comment by mattjj** [comment]
> Could you explain why you'd want a O(k^3) solution if O(k^2) solution exists and is implemented?

It's just as you say when talking about the O(k log k) option: in general, asymptotic analysis doesn't tell us what is faster for any concrete value of k. It could be that the O(k^2) solution is faster at all values of $k$, or it could be that the O(k^3) solution is faster for small k. 

The only bit of info we have is a paper saying (big caveat: IIRC) that their O(k^3) implementation was faster than the JAX code. There are many possible explanations. I just want to investigate that empirically.

**Comment by thisiscam** [comment]
I see! From what I gather from the paper, it’s not clear whether they’re using the O(k^3) solution. They do mention Faà di Bruno, but also note the use of specialized rules in their PyTorch Taylor mode: ![Image](https://github.com/user-attachments/assets/5bc63327-8dd0-4f93-a57a-7f8bea170b15)
That makes me think they might be doing something similar to jet:



They mention that their PyTorch Taylor mode is faster than jet. However, if I’m reading this correctly, the comparison is based on their technique for evaluating multiple K-jets in parallel. In other words, the benchmark seems to be between _collapsed Taylor + PyTorch_ v.s. _uncollapsed Taylor + JAX/jet_ ?


This is all guesses at this point without seeing their code.

**Comment by mattjj** [comment]
Good analysis, I agree!

**Comment by pnkraemer** [comment]
Hi! As a [heavy user of jet](https://pnkraemer.github.io/probdiffeq/api_docs/taylor/), I really like this change! Thanks for implementing it :) 

I also suspect that numerical robustness is going to improve by normalising the Taylor coefficients: high-order derivatives of ODE solutions can be really large, and scaling them down with a factorial seems like a good idea. 

Btw, recently jet became much quicker ([this benchmark looks different compared to a few months ago](https://pnkraemer.github.io/probdiffeq/benchmarks/taylor_node/plot/)). I don't know whether it's due to @thisiscam's change or something else, but it's much appreciated either way :)

**Comment by thisiscam** [comment]
@pnkraemer Thanks! Yes, I also observed accuracy improvements. 

For your use case, if you ever need to JIT compile with large order (as seen in your plots, n=30), I have an almost drop-in replacement of jet that operates with jax.scan instead of Python for loops. That way, compilation time should not scale significantly wrt n.

I can share that code with you to see if that is beneficial in any way --- it's still early WIP and I don't yet know how to PR that into jax yet.

**Comment by pnkraemer** [comment]
This sounds pretty neat! Happy to have a look at it and see whether it speeds things up on my side. 

That said, the extremely high orders (n=30) are mainly for demoing trends in this specific benchmark. In reality I rarely go over n=8 or so as I've not found a compelling use-case for ultra-high-order ODE solvers just yet :)
